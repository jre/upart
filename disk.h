#ifndef HDR_UPART_DISK
#define HDR_UPART_DISK

struct up_opts;

struct up_disk
{
    char   *upd_name;           /* disk name supplied by the user */
    char   *upd_path;           /* path to opened device node or plain file */
    int     upd_fd;             /* descriptor for disk device */
    int     upd_sectsize;       /* size of a sector in bytes */
    int64_t upd_cyls;           /* total number of cylinders */
    int64_t upd_heads;          /* number of tracks per cylinder */
    int64_t upd_sects;          /* number of sectors per track */
    int64_t upd_size;           /* total number of sects */
};

/* Open the disk device read-only and get drive params */
struct up_disk *up_disk_open(const char *path);

/* Close disk */
void up_disk_close(struct up_disk *disk);

void up_disk_dump(const struct up_disk *disk, void *_stream,
                  const struct up_opts *opt);

#endif /* HDR_UPART_DISK */
