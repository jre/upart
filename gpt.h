#ifndef HDR_UPART_GPT
#define HDR_UPART_GPT

/* register GPT partition map type */
void up_gpt_register(void);

#endif /* HDR_UPART_GPT */
