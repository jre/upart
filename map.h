#ifndef HDR_UPART_MAP
#define HDR_UPART_MAP

#include "bsdqueue.h"

struct up_map;
struct up_part;

#define UP_PART_EMPTY           (1<<0) /* empty or deleted */
#define UP_PART_OOB             (1<<1) /* out of bounds */

#define UP_PART_IS_BAD(flags)   ((UP_PART_EMPTY|UP_PART_OOB) & (flags))

SIMPLEQ_HEAD(up_map_list, up_map);
SIMPLEQ_HEAD(up_part_list, up_part);

struct up_part
{
    int64_t             start;
    int64_t             size;
    int                 flags;
    int                 depth;
    void *              priv;
    struct up_map      *map;
    struct up_part     *parent;
    struct up_part_list subpart;
    struct up_map_list  submap;
    SIMPLEQ_ENTRY(up_part) link;
};

enum up_map_type
{
    UP_MAP_NONE = 0,
    UP_MAP_MBR,
    //UP_MAP_BSD,
    UP_MAP_TYPE_COUNT
};

struct up_map
{
    struct up_disk     *disk;
    enum up_map_type    type;
    int64_t             start;
    int64_t             size;
    int                 depth;
    void *              priv;
    struct up_part     *parent;
    struct up_part_list list;
    SIMPLEQ_ENTRY(up_map) link;
};

void up_map_register(enum up_map_type type, const char *typestr,
                     /* check if map exists and allocate private data */
                     int (*load)(struct up_disk *, int64_t, int64_t, void **),
                     /* add partitions, misc setup not done in load */
                     int (*setup)(struct up_map *),
                     /* copy map header line into string */
                     int (*getinfo)(const struct up_map *, char *, int),
                     /* copy part index into string */
                     int (*getindex)(const struct up_part *, char *, int),
                     /* copy extra verbose info into string */
                     int (*getextra)(const struct up_part *, int, char *, int),
                     /* print hex dump of raw partition data to stream */
                     void (*dump)(const struct up_map *, void *),
                     /* free map private data, map may be NULL */
                     void (*freeprivmap)(struct up_map *, void *),
                     /* free part private data, part may be NULL */
                     void (*freeprivpart)(struct up_part *, void *));

int up_map_loadall(struct up_disk *disk);
void up_map_freeall(struct up_disk *disk);

int up_map_load(struct up_disk *disk, struct up_part *parent, int64_t start,
                int64_t size, enum up_map_type type, struct up_map **mapret);
struct up_part *up_map_add(struct up_map *map, struct up_part *parent,
                           int64_t start, int64_t size, int flags, void *priv);

void up_map_free(struct up_map *map);
void up_map_freeprivmap_def(struct up_map *map, void *priv);
void up_map_freeprivpart_def(struct up_part *part, void *priv);

void up_map_print(const struct up_map *map, void *stream,
                  int verbose, int recurse);
void up_map_dump(const struct up_map *map, void *stream, int recurse);
void up_map_printall(const struct up_disk *disk, void *stream, int verbose);
void up_map_dumpall(const struct up_disk *disk, void *stream);

const struct up_part *up_map_first(const struct up_map *map);
const struct up_part *up_map_firstsub(const struct up_part *part);
const struct up_part *up_map_next(const struct up_part *part);
const struct up_map  *up_map_firstmap(const struct up_part *part);
const struct up_map  *up_map_nextmap(const struct up_map *map);

#endif /* HDR_UPART_MAP */
