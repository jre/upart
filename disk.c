#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "disk.h"
#include "img.h"
#include "map.h"
#include "os.h"
#include "util.h"

/* #define DEBUG_SECTOR_SAVE */

static int fixparams(struct up_disk *disk, const struct up_opts *opts);
static int fixparams_checkone(struct up_diskparams *disk);
static int sectcmp(struct up_disk_sectnode *left,
                   struct up_disk_sectnode *right);

RB_GENERATE_STATIC(up_disk_sectmap, up_disk_sectnode, link, sectcmp);

struct up_disk *
up_disk_open(const char *name, const struct up_opts *opts,
             int writable)
{
    struct up_disk *disk;
    struct up_img  *img;
    const char     *path;
    int             fd, res, plain;
    struct stat     sb;

    /* open device */
    fd = up_os_opendisk(name, &path, opts, writable);
    if(0 > fd)
    {
        if(UP_NOISY(opts->verbosity, QUIET))
            up_err("failed to open %s for reading: %s",
                   (path ? path : name), strerror(errno));
        return NULL;
    }

    /* is it a plain file? */
    if(0 == fstat(fd, &sb) && S_ISREG(sb.st_mode))
        plain = 1;
    else
        plain = 0;

    /* check if it's an image file */
    img = NULL;
    if(plain)
    {
        res = up_img_load(fd, name, opts, &img);
        if(0 > res)
        {
            close(fd);
            return NULL;
        }
        else if(0 == res)
            assert(NULL != img);
    }

    /* disk struct */
    disk = calloc(1, sizeof *disk);
    if(!disk)
    {
        perror("malloc");
        close(fd);
        return NULL;
    }
    disk->ud_flag_plainfile = plain;
    RB_INIT(&disk->upd_sectsused);
    disk->upd_sectsused_count = 0;
    disk->ud_name = strdup(name);
    if(!disk->ud_name)
    {
        perror("malloc");
        close(fd);
        goto fail;
    }
    disk->ud_path = strdup(path ? path : disk->ud_name);
    if(!disk->ud_path)
    {
        perror("malloc");
        goto fail;
    }

    if(NULL == img)
    {
        /* it's not an image, initalize the disk struct normally */
        disk->upd_fd = fd;

        /* try to get drive parameters from OS */
        up_os_getparams(fd, disk, opts);
    }
    else
    {
        /* it is an image, initalize the disk accordingly */
        close(fd);
        disk->upd_fd = -1;
        disk->upd_img = img;

        /* try to get drive parameters from image */
        if(0 > up_img_getparams(&disk->ud_params, img))
            goto fail;
    }

    /* try to fill in missing drive paramaters */
    if(0 > fixparams(disk, opts))
    {
        if(UP_NOISY(opts->verbosity, QUIET))
            up_err("failed to determine disk parameters for %s",
                   UP_DISK_PATH(disk));
        goto fail;
    }

    return disk;

  fail:
    up_disk_close(disk);
    return NULL;
}

int64_t
up_disk_read(const struct up_disk *disk, int64_t start, int64_t size,
             void *buf, size_t bufsize, int verbose)
{
    ssize_t res;

    /* validate and fixup arguments */
    assert(0 < bufsize && 0 < size && 0 <= start);
    bufsize -= bufsize % UP_DISK_1SECT(disk);
    if(bufsize / UP_DISK_1SECT(disk) < size)
        size = bufsize / UP_DISK_1SECT(disk);
    if(0 == bufsize || 0 == size)
        return 0;
    if(INT64_MAX / UP_DISK_1SECT(disk) < start)
    {
        errno = EINVAL;
        return -1;
    }

    /* if there's an image then read from it instead */
    if(UP_DISK_IS_IMG(disk))
        return up_img_read(disk->upd_img, start, size, buf, verbose);

    /* otherwise try to read from the disk */
    res = pread(disk->upd_fd, buf, size * UP_DISK_1SECT(disk),
                start * UP_DISK_1SECT(disk));
    if(0 > res)
    {
        if(UP_NOISY(verbose, QUIET))
            up_err("failed to read %s sector %"PRIu64": %s",
                   UP_DISK_PATH(disk), start, strerror(errno));
        return -1;
    }

    return res / UP_DISK_1SECT(disk);
}

const void *
up_disk_getsect(struct up_disk *disk, int64_t sect, int vrb)
{
    if(!disk->upd_buf)
    {
        disk->upd_buf = malloc(UP_DISK_1SECT(disk));
        if(!disk->upd_buf)
        {
            perror("malloc");
            return NULL;
        }
    }

    if(1 > up_disk_read(disk, sect, 1, disk->upd_buf, UP_DISK_1SECT(disk), vrb))
        return NULL;
    else
        return disk->upd_buf;
}

int
up_disk_check1sect(struct up_disk *disk, int64_t sect)
{
    return up_disk_checksectrange(disk, sect, 1);
}

int
up_disk_checksectrange(struct up_disk *disk, int64_t start, int64_t size)
{
    struct up_disk_sectnode key;

    assert(0 < size);
    memset(&key, 0, sizeof key);
    key.first = start;
    key.last  = start + size - 1;
    if(NULL == RB_FIND(up_disk_sectmap, &disk->upd_sectsused, &key))
    {
#ifdef DEBUG_SECTOR_SAVE
        printf("check %"PRId64"+%"PRId64": free\n", start, size);
#endif
        return 0;
    }
    else
    {
#ifdef DEBUG_SECTOR_SAVE
        printf("check %"PRId64"+%"PRId64": used\n", start, size);
#endif
        return 1;
    }
}

const void *
up_disk_save1sect(struct up_disk *disk, int64_t sect,
                  const struct up_map *ref, int tag, int verbose)
{
    return up_disk_savesectrange(disk, sect, 1, ref, tag, verbose);
}

const void *
up_disk_savesectrange(struct up_disk *disk, int64_t first, int64_t size,
                      const struct up_map *ref, int tag, int verbose)
{
    struct up_disk_sectnode *new;

    /* allocate data structure */
    assert(0 < size);
    new = calloc(1, sizeof *new);
    if(NULL == new)
    {
        perror("malloc");
        return NULL;
    }
    new->data  = NULL;
    new->first = first;
    new->last  = first + size - 1;
    new->ref   = ref;
    new->tag   = tag;
    new->data  = calloc(size, UP_DISK_1SECT(disk));
    if(!new->data)
    {
        perror("malloc");
        free(new);
        return NULL;
    }

    /* try to read the sectors from disk */
    if(size != up_disk_read(disk, first, size, new->data,
                            size * UP_DISK_1SECT(disk), verbose))
    {
        free(new->data);
        free(new);
        return NULL;
    }

    /* insert it in the tree if the sectors aren't marked as used */
    if(RB_INSERT(up_disk_sectmap, &disk->upd_sectsused, new))
    {
#ifdef DEBUG_SECTOR_SAVE
        printf("failed to mark %"PRId64"+%"PRId64" with %p, already used\n",
               first, size, ref);
#endif
        free(new->data);
        free(new);
        return NULL;
    }
#ifdef DEBUG_SECTOR_SAVE
    printf("mark %"PRId64"+%"PRId64" with %p\n", first, size, ref);
#endif
    disk->upd_sectsused_count += size;

    /* return the data */
    return new->data;
}

void
up_disk_sectsunref(struct up_disk *disk, const void *ref)
{
    struct up_disk_sectnode *ii, *inc;

    for(ii = RB_MIN(up_disk_sectmap, &disk->upd_sectsused); ii; ii = inc)
    {
        inc = RB_NEXT(up_disk_sectmap, &disk->upd_sectsused, ii);
        if(ref == ii->ref)
        {
#ifdef DEBUG_SECTOR_SAVE
            printf("unmark %"PRId64"+%"PRId64" with %p\n",
                   ii->first, ii->last - ii->first + 1, ii->ref);
#endif
            RB_REMOVE(up_disk_sectmap, &disk->upd_sectsused, ii);
            disk->upd_sectsused_count -= ii->last - ii->first + 1;
            assert(0 <= disk->upd_sectsused_count);
            free(ii->data);
            free(ii);
        }
    }
}

void
up_disk_close(struct up_disk *disk)
{
    if(!disk)
        return;
    up_map_freeall(disk);
    assert(RB_EMPTY(&disk->upd_sectsused));
    assert(0 == disk->upd_sectsused_count);
    if(0 <= disk->upd_fd)
        close(disk->upd_fd);
    up_img_free(disk->upd_img);
    free(disk->upd_buf);
    free(disk->ud_name);
    free(disk->ud_path);
    free(disk);
}

static int
fixparams(struct up_disk *disk, const struct up_opts *opts)
{
    struct stat sb;

    /* command-line options override any other values */
    if(0 < opts->params.ud_sectsize)
        disk->ud_params.ud_sectsize = opts->params.ud_sectsize;
    if(0 < opts->params.ud_cyls)
        disk->ud_params.ud_cyls = opts->params.ud_cyls;
    if(0 < opts->params.ud_heads)
        disk->ud_params.ud_heads = opts->params.ud_heads;
    if(0 < opts->params.ud_sects)
        disk->ud_params.ud_sects = opts->params.ud_sects;

    /* sector size defaults to 512 */
    if(0 >= disk->ud_params.ud_sectsize)
        disk->ud_params.ud_sectsize = 512;

    /* we can get the total size for a plain file */
    if(0 >= disk->ud_params.ud_size && !UP_DISK_IS_IMG(disk) &&
       UP_DISK_IS_FILE(disk) && 0 == stat(UP_DISK_PATH(disk), &sb))
        disk->ud_params.ud_size = sb.st_size / disk->ud_params.ud_sectsize;

    /* is that good enough? */
    if(0 == fixparams_checkone(&disk->ud_params))
        return 0;

    /* apparently not, try defaulting heads and sectors to 255 and 63 */
    if(0 >= disk->ud_params.ud_heads)
        disk->ud_params.ud_heads = 255;
    if(0 >= disk->ud_params.ud_sects)
        disk->ud_params.ud_sects = 63;

    /* ok, is it good enough now? */
    if(0 == fixparams_checkone(&disk->ud_params))
        return 0;

    /* guess not */
    return -1;
}

static int
fixparams_checkone(struct up_diskparams *disk)
{
    /* if we have all 4 then we're good */
    if(0 <  disk->ud_cyls  && 0 <  disk->ud_heads &&
       0 <  disk->ud_sects && 0 <  disk->ud_size )
        return 0;

    /* guess cyls from other three */
    if(0 >= disk->ud_cyls  && 0 <  disk->ud_heads &&
       0 <  disk->ud_sects && 0 <  disk->ud_size )
    {
        disk->ud_cyls = disk->ud_size / disk->ud_sects / disk->ud_heads;
        return 0;
    }

    /* guess heads from other three */
    if(0 <  disk->ud_cyls  && 0 >= disk->ud_heads &&
       0 <  disk->ud_sects && 0 <  disk->ud_size )
    {
        disk->ud_heads = disk->ud_size / disk->ud_sects / disk->ud_cyls;
        return 0;
    }

    /* guess sects from other three */
    if(0 <  disk->ud_cyls  && 0 <  disk->ud_heads &&
       0 >= disk->ud_sects && 0 <  disk->ud_size )
    {
        disk->ud_sects = disk->ud_size / disk->ud_heads / disk->ud_cyls;
        return 0;
    }

    /* guess size from other three */
    if(0 <  disk->ud_cyls  && 0 <  disk->ud_heads &&
       0 <  disk->ud_sects && 0 >= disk->ud_size )
    {
        disk->ud_size = disk->ud_sects * disk->ud_heads * disk->ud_cyls;
        return 0;
    }

    /* can only guess with one missing param */
    return -1;
}

void
up_disk_print(const struct up_disk *disk, void *_stream, int verbose)
{
    FILE *              stream = _stream;
    const char *        unit;
    float               size;

    size = up_fmtsize(UP_DISK_SIZEBYTES(disk), &unit);
    if(UP_NOISY(verbose, NORMAL))
        fprintf(stream, "%s: %.*f%s (%"PRId64" sectors of %d bytes)\n",
                UP_DISK_PATH(disk), UP_BESTDECIMAL(size), size, unit,
                UP_DISK_SIZESECTS(disk), UP_DISK_1SECT(disk));
    if(UP_NOISY(verbose, EXTRA))
        fprintf(stream,
                "    label:               %s\n"
                "    device path:         %s\n"
                "    sector size:         %d\n"
                "    total sectors:       %"PRId64"\n"
                "    total cylinders:     %"PRId64" (cylinders)\n"
                "    tracks per cylinder: %"PRId64" (heads)\n"
                "    sectors per track:   %"PRId64" (sectors)\n"
                "\n", UP_DISK_LABEL(disk),
                UP_DISK_PATH(disk), UP_DISK_1SECT(disk), UP_DISK_SIZESECTS(disk),
                UP_DISK_CYLS(disk), UP_DISK_HEADS(disk), UP_DISK_SPT(disk));
    if(UP_NOISY(verbose, NORMAL))
        fputc('\n', stream);
}

void
up_disk_dump(const struct up_disk *disk, void *stream)
{
    struct up_disk_sectnode *node;

    for(node = RB_MIN(up_disk_sectmap, &disk->upd_sectsused); node;
        node = RB_NEXT(up_disk_sectmap, &disk->upd_sectsused, node))
        up_map_dumpsect(node->ref, stream, node->first,
                        node->last - node->first + 1, node->data, node->tag);
}

void
up_disk_sectsiter(const struct up_disk *disk,
                  void (*func)(const struct up_disk *,
                               const struct up_disk_sectnode *, void *),
                  void *arg)
{
    struct up_disk_sectnode *node;

    for(node = RB_MIN(up_disk_sectmap, &disk->upd_sectsused); node;
        node = RB_NEXT(up_disk_sectmap, &disk->upd_sectsused, node))
        func(disk, node, arg);
}

const struct up_disk_sectnode *
up_disk_nthsect(const struct up_disk *disk, int off)
{
    struct up_disk_sectnode *node;

    assert(0 <= off);
    for(node = RB_MIN(up_disk_sectmap, &disk->upd_sectsused); node;
        node = RB_NEXT(up_disk_sectmap, &disk->upd_sectsused, node))
        if(0 == off--)
            return node;
    return NULL;
}

static int
sectcmp(struct up_disk_sectnode *left, struct up_disk_sectnode *right)
{
    if(left->last < right->first)
    {
        /*
        printf("cmp (%"PRId64"+%"PRId64") < (%"PRId64"+%"PRId64")\n",
               left->first, left->last - left->first + 1,
               right->first, right->last - right->first + 1);
        */
        return -1;
    }
    else if(left->first > right->last)
    {
        /*
        printf("cmp (%"PRId64"+%"PRId64") > (%"PRId64"+%"PRId64")\n",
               left->first, left->last - left->first + 1,
               right->first, right->last - right->first + 1);
        */
        return 1;
    }
    else
    {
        /*
        printf("cmp (%"PRId64"+%"PRId64") = (%"PRId64"+%"PRId64")\n",
               left->first, left->last - left->first + 1,
               right->first, right->last - right->first + 1);
        */
        return 0;
    }
}
