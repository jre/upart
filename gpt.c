#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "crc32.h"
#include "disk.h"
#include "gpt.h"
#include "map.h"
#include "util.h"

#define GPT_SIZE                0x5c
#define GPT_PRIOFF(st, sz)      ((st) + UINT64_C(1))
#define GPT_PRISIZE(st, sz)     ((sz) - UINT64_C(1))
#define GPT_SECOFF(st, sz)      ((st) + (sz) - UINT64_C(1))
#define GPT_SECSIZE(st, sz)     (UINT64_C(1))
#define GPT_OFFSEC              -1
#define GPT_MAGIC               UINT64_C(0x5452415020494645)
#define GPT_REVISION            0x10000
#define GPT_GUID_FMT \
    "%08x-%04x-%04x-%02x%02x-%02x%02x%02x%02x%02x%02x"
#define GPT_GUID_FMT_ARGS(guid) \
    UP_LETOH32((guid)->data1), UP_LETOH16((guid)->data2), \
    UP_LETOH16((guid)->data3), (guid)->data4[0], (guid)->data4[1], \
    (guid)->data4[2], (guid)->data4[3], (guid)->data4[4], (guid)->data4[5], \
    (guid)->data4[6], (guid)->data4[7]
#define GPT_GUID_DATA4_SIZE     8
#define GPT_PART_SIZE           0x80
#define GPT_NAME_SIZE           0x48

struct up_guid_p
{
    uint32_t            data1;
    uint16_t            data2;
    uint16_t            data3;
    uint8_t             data4[GPT_GUID_DATA4_SIZE];
} __attribute__((packed));

struct up_gpt_p
{
    uint64_t            magic;
    uint32_t            revision;
    uint32_t            size;
    uint32_t            gptcrc;
    uint32_t            unused;
    uint64_t            gpt1sect;
    uint64_t            gpt2sect;
    uint64_t            firstsect;
    uint64_t            lastsect;
    struct up_guid_p    guid;
    uint64_t            partsect;
    uint32_t            maxpart;
    uint32_t            partsize;
    uint32_t            partcrc;
} __attribute__((packed));

struct up_gptpart_p
{
    struct up_guid_p    type;
    struct up_guid_p    guid;
    uint64_t            start;
    uint64_t            end;
    uint64_t            flags;
    char                name[GPT_NAME_SIZE];
} __attribute__((packed));

struct up_gptpart
{
    struct up_gptpart_p part;
    int                 index;
};

struct up_gpt
{
    struct up_gpt_p     gpt;
    int                 partitions;
};

static int gpt_load(struct up_disk *disk, const struct up_part *parent,
                    void **priv, const struct up_opts *opts);
static int gpt_setup(struct up_map *map, const struct up_opts *opts);
static int gpt_getinfo(const struct up_map *part, int verbose,
                       char *buf, int size);
static int gpt_getindex(const struct up_part *part, char *buf, int size);
static int gpt_getextra(const struct up_part *part, int verbose,
                        char *buf, int size);
static int gpt_findhdr(struct up_disk *disk, int64_t start, int64_t size,
                       struct up_gpt_p *gpt);
static int gpt_readhdr(struct up_disk *disk, int64_t start, int64_t size,
                       const struct up_gpt_p **gpt);
static int gpt_checkcrc(struct up_gpt_p *gpt);
static const char *gpt_typename(const struct up_guid_p *guid);

void
up_gpt_register(void)
{
    up_map_register(UP_MAP_GPT,
                    "EFI GPT",
                    0,
                    gpt_load,
                    gpt_setup,
                    gpt_getinfo,
                    gpt_getindex,
                    gpt_getextra,
                    NULL,
                    up_map_freeprivmap_def,
                    up_map_freeprivpart_def);
}

int
gpt_load(struct up_disk *disk, const struct up_part *parent, void **priv,
         const struct up_opts *opts)
{
    struct up_gpt_p pk;
    int             res;
    struct up_gpt  *gpt;

    assert(GPT_SIZE == sizeof pk &&
           GPT_PART_SIZE == sizeof(struct up_gptpart_p));
    *priv = NULL;

    /* try to load either the primary or secondary gpt headers, and
       check the magic and crc */
    res = gpt_findhdr(disk, parent->start, parent->size, &pk);
    if(0 >= res)
        return res;

    /* check revision */
    if(GPT_REVISION != UP_LETOH32(pk.revision))
    {
        fprintf(stderr, "ignoring gpt with unknown revision %u.%u\n",
                (UP_LETOH32(pk.revision) >> 16) & 0xffff,
                UP_LETOH32(pk.revision) & 0xffff);
        return 0;
    }

    /* XXX validate other fields */

    /* create map private struct */
    gpt = calloc(1, sizeof *gpt);
    if(!gpt)
    {
        perror("malloc");
        return -1;
    }
    gpt->gpt = pk;
    *priv = gpt;

    return 1;
}

static int
gpt_setup(struct up_map *map, const struct up_opts *opts)
{
    struct up_gpt              *priv = map->priv;
    struct up_gpt_p            *gpt = &priv->gpt;
    struct up_gptpart          *part;
    const struct up_gptpart_p  *pk;
    uint64_t                    partsects, partbytes;
    const uint8_t              *data1, *data2;

    /* calculate partition buffer size */
    partbytes = UP_LETOH32(gpt->maxpart) * GPT_PART_SIZE;
    partsects = partbytes / map->disk->upd_sectsize;

    /* save sectors from primary and secondary maps */
    data1 = up_disk_savesectrange(map->disk, GPT_PRIOFF(map->start, map->size),
                                  1 + partsects, map, 0);
    data2 = up_disk_savesectrange(map->disk, GPT_SECOFF(map->start, map->size)
                                  - partsects, 1 + partsects, map, 0);
    if(!data1 || !data2)
        return -1;

    /* verify the crc */
    if(UP_LETOH32(gpt->partcrc) !=
       (up_crc32(data1 + map->disk->upd_sectsize,  partbytes, ~0) ^ ~0))
    {
        fprintf(stderr, "bad gpt partition crc\n");
        return 0;
    }

    /* walk through the partition buffer and add all partitions found */
    pk = (const struct up_gptpart_p *)(data1 + map->disk->upd_sectsize);
    while((const uint8_t *)pk + sizeof *pk
          <= data1 + map->disk->upd_sectsize * partsects)
    {
        part = calloc(1, sizeof *part);
        if(!part)
        {
            perror("malloc");
            return -1;
        }
        part->part = *pk;
        part->index = priv->partitions;
        if(!up_map_add(map, UP_LETOH64(pk->start), UP_LETOH64(pk->end)
                       - UP_LETOH64(pk->start), 0, part))
        {
            free(part);
            return -1;
        }
        priv->partitions++;
        pk++;
    }

    return 1;
}

static int
gpt_getinfo(const struct up_map *map, int verbose, char *buf, int size)
{
    const struct up_gpt *gpt = map->priv;

    if(UP_NOISY(verbose, EXTRA))
        return snprintf(buf, size, "EFI GPT partition table in sectors %"PRId64" "
                    "and %"PRId64" of %s:\n"
                    "  size:                 %u\n"
                    "  primary gpt sector:   %"PRIu64"\n"
                    "  backup gpt sector:    %"PRIu64"\n"
                    "  first data sector:    %"PRIu64"\n"
                    "  last data sector:     %"PRIu64"\n"
                    "  guid:                 "GPT_GUID_FMT"\n"
                    "  partition sector:     %"PRIu64"\n"
                    "  max partitions:       %u\n"
                    "  partition size:       %u\n"
                    "\n",
                    GPT_PRIOFF(map->start, map->size),
                    GPT_SECOFF(map->start, map->size), map->disk->upd_name,
                    UP_LETOH32(gpt->gpt.size),
                    UP_LETOH64(gpt->gpt.gpt1sect),
                    UP_LETOH64(gpt->gpt.gpt2sect),
                    UP_LETOH64(gpt->gpt.firstsect),
                    UP_LETOH64(gpt->gpt.lastsect),
                    GPT_GUID_FMT_ARGS(&gpt->gpt.guid),
                    UP_LETOH64(gpt->gpt.partsect),
                    UP_LETOH32(gpt->gpt.maxpart),
                    UP_LETOH32(gpt->gpt.partsize));
    else if(UP_NOISY(verbose, NORMAL))
        return snprintf(buf, size, "EFI GPT partition table in sectors %"PRId64
                        " and %"PRId64" of %s:",
                        GPT_PRIOFF(map->start, map->size),
                        GPT_SECOFF(map->start, map->size), map->disk->upd_name);
    else
        return 0;
}

static int
gpt_getindex(const struct up_part *part, char *buf, int size)
{
    const struct up_gptpart *priv = part->priv;

    return snprintf(buf, size, "%d", priv->index + 1);
}

static int
gpt_getextra(const struct up_part *part, int verbose, char *buf, int size)
{
    const struct up_gptpart    *priv;
    const char                 *label;

    if(!UP_NOISY(verbose, NORMAL))
        return 0;

    if(!part)
    {
        if(UP_NOISY(verbose, EXTRA))
            return snprintf(buf, size, "%-36s Type", "GUID");
        else
            return snprintf(buf, size, "Type");
    }

    priv = part->priv;
    label = gpt_typename(&priv->part.type);

    if(UP_NOISY(verbose, EXTRA))
        return snprintf(buf, size, GPT_GUID_FMT " " GPT_GUID_FMT " %s",
                        GPT_GUID_FMT_ARGS(&priv->part.guid),
                        GPT_GUID_FMT_ARGS(&priv->part.type),
                        (label ? label : ""));
    else
    {
        if(label)
            return snprintf(buf, size, "%s", label);
        else
            return snprintf(buf, size, GPT_GUID_FMT,
                            GPT_GUID_FMT_ARGS(&priv->part.type));
    }
}

static int
gpt_findhdr(struct up_disk *disk, int64_t start, int64_t size,
            struct up_gpt_p *gpt)
{
    const struct up_gpt_p  *buf;
    int                     res, badcrc;

    badcrc = 0;

    res = gpt_readhdr(disk, GPT_PRIOFF(start, size),
                      GPT_PRISIZE(start, size), &buf);
    if(0 >= res)
        return res;
    if(GPT_MAGIC == UP_LETOH64(buf->magic))
    {
        *gpt = *buf;
        if(gpt_checkcrc(gpt))
            return 1;
        fprintf(stderr, "bad crc on primary gpt in sector %"PRId64"\n",
                GPT_PRIOFF(start, size));
        badcrc = 1;
    }

    res = gpt_readhdr(disk, GPT_SECOFF(start, size),
                      GPT_SECSIZE(start, size), &buf);
    if(0 >= res)
        return res;
    if(GPT_MAGIC == UP_LETOH64(buf->magic))
    {
        *gpt = *buf;
        if(gpt_checkcrc(gpt))
            return 1;
        fprintf(stderr, "bad crc on secondary gpt in sector %"PRId64"\n",
                GPT_SECOFF(start, size));
        badcrc = 1;
    }

    return (badcrc ? -1 : 0);
}

static int
gpt_readhdr(struct up_disk *disk, int64_t start, int64_t size,
            const struct up_gpt_p **gpt)
{
    const void *buf;

    *gpt = NULL;

    if(0 >= size || sizeof *gpt > disk->upd_sectsize)
        return 0;

    if(up_disk_check1sect(disk, start))
        return 0;
    buf = up_disk_getsect(disk, start);
    if(!buf)
        return -1;
    *gpt = buf;

    return 1;
}

static int
gpt_checkcrc(struct up_gpt_p *gpt)
{
    uint32_t save, crc;

    save        = gpt->gptcrc;
    gpt->gptcrc = 0;
    crc         = ~0 ^ up_crc32(gpt, sizeof *gpt, ~0);
    gpt->gptcrc = save;

    return UP_LETOH32(save) == crc;
}

static const uint8_t gpt_data4_apple[] = "\xaa\x11\x00\x30\x65\x43\xec\xac";
static const uint8_t gpt_data4_sun[]   = "\x99\xA6\x08\x00\x20\x73\x66\x31";
static struct
{
    uint32_t data1;
    uint16_t data2;
    uint16_t data3;
    const uint8_t *data4;
    const char *name;
} gpt_labels[] =
{
    {0, 0, 0, (const uint8_t*)"\0\0\0\0\0\0\0\0", "unused"},
    {0xc12a7328, 0xf81f, 0x11d2, (const uint8_t*)"\xba\x4b\x00\xa0\xc9\x3e\xc9\x3b", "EFI System Partition"},
    {0x48465300, 0, 0x11aa, gpt_data4_apple, "Apple HFS+"},
    {0x55465300, 0, 0x11aa, gpt_data4_apple, "Apple UFS"},
    {0xebd0a0a2, 0xb9e5, 0x4433, (const uint8_t*)"\x87\xc0\x68\xb6\xb7\x26\x99\xc7", "Microsoft Data"},
    {0x6A898CC3, 0x1dd2, 0x11b2, gpt_data4_sun, "Solaris /usr or Apple ZFS"},
};

static const char *
gpt_typename(const struct up_guid_p *guid)
{
    uint32_t data1;
    uint16_t data2, data3;
    int ii;

    data1 = UP_LETOH32(guid->data1);
    data2 = UP_LETOH16(guid->data2);
    data3 = UP_LETOH16(guid->data3);
    for(ii = 0; sizeof(gpt_labels) / sizeof(gpt_labels[0]) > ii; ii++)
    {
        if(data1 == gpt_labels[ii].data1 &&
           data2 == gpt_labels[ii].data2 &&
           data3 == gpt_labels[ii].data3 &&
           0 == memcmp(guid->data4, gpt_labels[ii].data4, GPT_GUID_DATA4_SIZE))
            return gpt_labels[ii].name;
    }

    return NULL;
}
