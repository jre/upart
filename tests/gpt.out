gpt.img: 960MB (1966080 sectors of 512 bytes)

EFI GPT partition table in sectors 1 and 1966079 of gpt.img:
                 Start            Size Type
1:                  40          491519 Apple HFS+
2:              491560          491519 Microsoft Data
3:              983080          491519 Apple UFS
4:             1474600          491439 Solaris /usr or Apple ZFS
MBR partition table in sector 0 of gpt.img:
                 Start            Size A Type
0:                   1         1966079   EFI GPT (0xee)
